### ISSUE TITLE
Module full backward hook does not work (perhaps because input does not require grad)

### ISSUE THREAD

**Issue by bluenote10** [pytorch/pytorch#117925]
### 🐛 Describe the bug

I'm trying to register a full backward hook but it does not get called. A minimal example would be:

```py
import torch

Gradients = torch.Tensor | tuple[torch.Tensor, ...]

class MyModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.some_param = torch.nn.Parameter(torch.tensor(1.0))
        self.register_full_backward_hook(self.full_backward_hook)

    @staticmethod
    def full_backward_hook(
        module: torch.nn.Module,
        grad_in: Gradients,
        grad_out: Gradients,
    ) -> None:
        print("full_backward_hook called with", module, grad_in, grad_out)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return x * self.some_param

model = MyModule()

output = model.forward(torch.tensor(5))
loss = output.sum()

print("model.some_param.grad:", model.some_param.grad)

print("calling backward")
loss.backward()

print("model.some_param.grad:", model.some_param.grad)
```

The actual output of this snippet is:
```
model.some_param.grad: None
calling backward
model.some_param.grad: tensor(5.)
```

The expected output would be:
```
model.some_param.grad: None
calling backward
full_backward_hook called with ...
model.some_param.grad: tensor(5.)
```

This is related to the old issue https://github.com/pytorch/pytorch/issues/56380 which has been fixed in PR https://github.com/pytorch/pytorch/pull/56693. Prior to that PR the issue was that full backward hooks did not fire if the inputs did not require grad. The PR describes the change as:

> This changes the behavior of full backward hooks as they will now fire properly even if no input to the Module require gradients.

So this may be an regression.

### Versions

<details>

<summary>Detailed versions</summary>

```
Collecting environment information...
PyTorch version: 2.1.2+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-91-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 980
Nvidia driver version: 535.146.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      39 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             4
On-line CPU(s) list:                0-3
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Core(TM) i5-4670 CPU @ 3.40GHz
CPU family:                         6
Model:                              60
Thread(s) per core:                 1
Core(s) per socket:                 4
Socket(s):                          1
Stepping:                           3
CPU max MHz:                        3800,0000
CPU min MHz:                        800,0000
BogoMIPS:                           6799.77
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid xsaveopt dtherm ida arat pln pts md_clear flush_l1d
Virtualization:                     VT-x
L1d cache:                          128 KiB (4 instances)
L1i cache:                          128 KiB (4 instances)
L2 cache:                           1 MiB (4 instances)
L3 cache:                           6 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-3
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX conditional cache flushes, SMT disabled
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT disabled
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Unknown: No mitigations
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Mitigation; Microcode
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] ddsp-pytorch==0.0.1
[pip3] flake8==7.0.0
[pip3] mypy==1.8.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.3
[pip3] torch==2.1.2
[pip3] torchaudio==2.1.2
[pip3] torchcrepe==0.0.22
[pip3] torchinfo==1.8.0
[pip3] torchviz==0.0.2
[pip3] triton==2.1.0
[conda] Could not collect
```

</details>

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by tringwald** [comment]
From the [docs of nn.Module.forward](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.forward):
>Although the recipe for forward pass needs to be defined within this function, one should call the [Module](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module) instance afterwards instead of this since the former takes care of running the registered hooks while the latter silently ignores them.

When replacing the direct `.forward()` invocation with `model(torch.tensor(5))` it works perfectly fine:
```
model.some_param.grad: None
calling backward
full_backward_hook called with MyModule() (None,) (tensor(1.),)
model.some_param.grad: tensor(5.)
```

**Comment by bluenote10** [comment]
> When replacing the direct .forward() invocation with model(torch.tensor(5)) it works perfectly fine

Ah, sorry for my mistake!

In this case I would say everything works as expected and this issue can probably be closed.
